ethereum/statetransition/src/main/java/tech/pegasys/artemis/statetransition/util/BeaconStateUtil.java

```diff
@@ -157,85 +157,92 @@ public static double calc_total_balance(BeaconState state) {
-    // TODO: convert these values to UnsignedLong
-    long epoch = slot_to_epoch(UnsignedLong.valueOf(slot)).longValue();
-    long current_epoch = get_current_epoch(state).longValue();
-    long previous_epoch =
-        (current_epoch > Constants.GENESIS_EPOCH) ? current_epoch - 1 : current_epoch;
-    long next_epoch = current_epoch + 1;
+      BeaconState state, UnsignedLong slot, boolean registry_change) {
+    UnsignedLong epoch = slot_to_epoch(slot);
+    UnsignedLong current_epoch = get_current_epoch(state);
+    UnsignedLong previous_epoch = get_previous_epoch(state);
+    UnsignedLong next_epoch = get_next_epoch(state);
 
-    assertTrue(previous_epoch <= epoch && epoch <= next_epoch);
+    assertTrue(previous_epoch.compareTo(epoch) <= 0 && epoch.compareTo(next_epoch) <= 0);
 
-    long committees_per_epoch = 0;
+    UnsignedLong committees_per_epoch = UnsignedLong.ZERO;
     Bytes32 seed = Bytes32.ZERO;
-    long shuffling_epoch = 0;
-    long shuffling_start_shard = 0;
+    UnsignedLong shuffling_epoch = UnsignedLong.ZERO;
+    UnsignedLong shuffling_start_shard = UnsignedLong.ZERO;
 
-    if (epoch == previous_epoch) {
+    if (epoch.compareTo(previous_epoch) == 0) {
 
       committees_per_epoch = get_previous_epoch_committee_count(state);
       seed = state.getPrevious_epoch_seed();
-      // TODO: convert these values to UnsignedLong
-      shuffling_epoch = state.getPrevious_calculation_epoch().longValue();
-      shuffling_start_shard = state.getPrevious_epoch_start_shard().longValue();
+      shuffling_epoch = state.getPrevious_calculation_epoch();
+      shuffling_start_shard = state.getPrevious_epoch_start_shard();
 
-    } else if (epoch == current_epoch) {
-      // TODO: convert these values to UnsignedLong
+    } else if (epoch.compareTo(current_epoch) == 0) {
       committees_per_epoch = get_current_epoch_committee_count(state);
       seed = state.getCurrent_epoch_seed();
-      shuffling_epoch = state.getCurrent_calculation_epoch().longValue();
-      shuffling_start_shard = state.getCurrent_epoch_start_shard().longValue();
+      shuffling_epoch = state.getCurrent_calculation_epoch();
+      shuffling_start_shard = state.getCurrent_epoch_start_shard();
 
-    } else if (epoch == next_epoch) {
-      // TODO: convert these values to UnsignedLong
-      long current_committees_per_epoch = get_current_epoch_committee_count(state);
+    } else if (epoch.compareTo(next_epoch) == 0) {
+      UnsignedLong current_committees_per_epoch = get_current_epoch_committee_count(state);
       committees_per_epoch = get_next_epoch_committee_count(state);
       shuffling_epoch = next_epoch;
-      long epochs_since_last_registry_update =
-          current_epoch - state.getValidator_registry_update_epoch().longValue();
-      // TODO: convert these values to UnsignedLong
+      UnsignedLong epochs_since_last_registry_update =
+          current_epoch.minus(state.getValidator_registry_update_epoch());
```

Comment:
It's annoying, but everywhere we do `UnsignedLong#minus`, we probably need bounds checking to make sure it doesn't underflow. I'm sure this is a pervasive problem, so just write up an issue and a todo.

Reply:
ya, but we need to implement it in a way so that we don't litter the code with it.  more than annoying, it creates a maintenance problem unless it is fixed in a better way